### PR TITLE
Add dynamic (duck) type resolution to Painless static types

### PR DESCRIPTION
**Pull request by jdconrad**
This change adds dynamic (duck) type resolution to Painless static types using an annotation ( at dynamic_type ) to control which static types are allowed to be dynamically invoked.  This annotation does not chain so any sub classes that also require dynamic type resolution must be annotated as well.

Example:
```
Java:
class A {
}

class B {
   String duck() { returns "duck"; }
}

allow list:
class A @dynamic_type {
}

class B {
  ()
  String duck()
}

Painless:
A a = new B();
a.duck();
```

Normally, this would fail as `a` is statically typed, so the method duck would not be found. With dynamic type resolution, this will succeed by seeing that we have a possible duck method on a sub class and binding the method at runtime instead.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java [hunk 1]
```diff
@@ -2904,34 +2943,12 @@ public void visitCall(ECall userCallNode, SemanticScope semanticScope) {
             }
 
             TargetType targetType = userCallNode.isNullSafe() ? null : semanticScope.getDecoration(userCallNode, TargetType.class);
-            // TODO: remove ZonedDateTime exception when JodaCompatibleDateTime is removed
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java:2907]: > JodaCompatibleDateTime

was this missed in https://github.com/elastic/elasticsearch/pull/78417?

**Review by stu-elastic — APPROVED** [review]
very clean

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java:2907]: Yes! I searched for the class usages as opposed to just the name.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review!

**Pull request merged**